File: apps/go/internal/repository/shipping.go
package repository

import (
	"context"

	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/ryansuhartanto/koda-b8-backend/apps/go/internal/model"
)

func ShippingMethods(ctx context.Context, pool *pgxpool.Pool) ([]model.ShippingMethod, error) {
	rows, err := pool.Query(ctx,
		`SELECT
			id,
			name,
			cost_idr
		FROM shipping_methods
		WHERE deleted_at IS NULL
		ORDER BY cost_idr, id`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	methods := []model.ShippingMethod{}

	for rows.Next() {
		var method model.ShippingMethod

		if err := rows.Scan(
			&method.ID,
			&method.Name,
			&method.CostIdr,
		); err != nil {
			return nil, err
		}

		methods = append(methods, method)
	}

	return methods, rows.Err()
}

File: apps/js/src/routes/shipping.ts
import { Router } from "express";

import { pool } from "#/lib/db";
import { problem } from "#/lib/problem";
import type { ShippingMethod } from "#/model/shipping";

export const router: Router = Router();

/**
 * @openapi
 * components:
 *   schemas:
 *     ShippingMethod:
 *       type: object
 *       properties:
 *         id: { type: integer }
 *         name: { type: string }
 *         cost_idr: { type: integer }
 *       required: [cost_idr, id, name]
 *
 * /shipping-methods:
 *   get:
 *     summary: List shipping methods and their cost
 *     tags: [shipping]
 *     responses:
 *       "200":
 *         description: OK
 *         content:
 *           application/json:
 *             schema:
 *               { type: array, items: { $ref: "#/components/schemas/ShippingMethod" } }
 *       "500":
 *         description: Internal error
 *         content:
 *           application/json:
 *             schema: { $ref: "#/components/schemas/Problem" }
 */
router.get("/shipping-methods", async (_req, res) => {
	try {
		const { rows } = await pool.query<ShippingMethod>(
			`SELECT
				id,
				name,
				cost_idr
			FROM shipping_methods
			WHERE deleted_at IS NULL
			ORDER BY cost_idr, id`,
		);

		const methods: ShippingMethod[] = rows;

		res.json(methods);
	} catch (error) {
		problem(res, 500, error);
	}
});
